x/incentives/keeper/autostaking.go

```diff
@@ -0,0 +1,123 @@
+	k.IterateAutoStaking(ctx, func(index int64, autostaking types.AutoStaking) (stop bool) {
+		autostakings = append(autostakings, autostaking)
+		return false
+	})
+	return autostakings
+}
+
+// SetAutostaking set the autostaking configuration into the store
+func (k Keeper) SetAutostaking(ctx sdk.Context, autostaking *types.AutoStaking) error {
+	store := ctx.KVStore(k.storeKey)
+	bz, err := proto.Marshal(autostaking)
+	if err != nil {
+		return err
+	}
+	store.Set(autostakingStoreKey(autostaking.Address), bz)
+	return nil
+}
+
+// GetAutostakingByAddress regurns autostaking config by address
+func (k Keeper) GetAutostakingByAddress(ctx sdk.Context, address string) *types.AutoStaking {
+	autostaking := types.AutoStaking{}
+	store := ctx.KVStore(k.storeKey)
+	autostakingKey := autostakingStoreKey(address)
+	if !store.Has(autostakingKey) {
+		return nil
+	}
+	bz := store.Get(autostakingKey)
+	proto.Unmarshal(bz, &autostaking)
+	return &autostaking
+}
+
+func (k Keeper) AutostakeRewards(ctx sdk.Context, owner sdk.AccAddress, distrCoins sdk.Coins) error {
+	params := k.GetParams(ctx)
+	bondDenom := k.sk.BondDenom(ctx)
+	bondDenomAmt := distrCoins.AmountOf(bondDenom)
+	// TODO: later should use user's manually configured delegation rate, for now, uses simple rate defined as param by governance
+	autoDelegationAmt := bondDenomAmt.ToDec().Mul(params.MinAutostakingRate).RoundInt()
+	if !autoDelegationAmt.IsPositive() {
+		return nil
+	}
+	// auto delegate when can
+	autostaking := k.GetAutostakingByAddress(ctx, owner.String())
+
+	autostaked := false
+	if autostaking != nil {
+		valAddr, err := sdk.ValAddressFromBech32(autostaking.AutostakingValidator)
+		if err != nil {
+			return err
+		}
+
+		validator, found := k.sk.GetValidator(ctx, valAddr)
+		if found {
+			// NOTE: source funds are always unbonded and the param for `tokenSrc` is `stakingtypes.Unbonded`
+			// Param for `subtractAccount` is `true` to substract delegating coins from owner's account.
+			_, err = k.sk.Delegate(ctx, owner, autoDelegationAmt, stakingtypes.Unbonded, validator, true)
+			if err != nil {
+				return err
+			}
+			autostaked = true
+		}
+	}
+	if !autostaked { // lock tokens forcefully - TODO: if lock tokens on every epoch, lots of locks will appear
+		// if there's a bond denom lock of 2 weeks already just add more tokens there
+		defaultLockDuration := time.Hour * 24 * 7 * 2
+		locks := k.lk.GetAccountLockedDurationDenom(ctx, owner, bondDenom, defaultLockDuration)
+		if len(locks) > 0 { // if same duration + denom lock already for the account, just add it there
+			_, err := k.lk.AddTokensToLock(ctx, owner, locks[0].ID, sdk.Coins{sdk.NewCoin(bondDenom, autoDelegationAmt)})
+			if err != nil {
+				return err
+			}
+		} else { // create a new lock if this lock is not available
+			_, err := k.lk.LockTokens(ctx, owner, sdk.Coins{sdk.NewCoin(bondDenom, autoDelegationAmt)}, time.Hour*24*7*2)
+			if err != nil {
+				return err
+			}
+		}
```

Comment:
Hmm, it will be confusing if a user need to provide lockID when creating a new one. Also the underlying functionality is different, creating a new lock is doing different stuff with Adding tokens to existing lock.

Reply:
Oh I see, you're totally right! What if it was `k.lk.AddTokensToLockedDurationDenom` that did all this logic?